tests/test_override_loop.py

```diff
@@ -133,6 +135,69 @@ async def blah():
         asyncio.get_event_loop().run_until_complete(info["coro"])
 
 
+it "can shutdown async gens":
+    info1 = []
+    info2 = []
+
+    original = asyncio.get_event_loop()
+
+    async def my_generator(info):
+        try:
+            info.append(1)
+            yield
+            info.append(2)
+            yield
+            info.append(3)
+        except asyncio.CancelledError:
+            info.append("cancelled")
+            raise
+        finally:
+            info.append(("done", __import__("sys").exc_info()[0]))
+
+    # Test that the outside loop isn't affected by the inside loop
+    outside_gen = my_generator(info1)
+
+    async def outside1():
+        await outside_gen.__anext__()
+        await outside_gen.__anext__()
+
+    original.run_until_complete(outside1())
+    assert info1 == [1, 2]
+
+    with OverrideLoop(new_loop=True) as custom_loop:
+        assert info2 == []
+
+        async def doit():
+            ag = my_generator(info2)
```

Comment:
OK, I tried this out myself and it's a a little complicated, but here's an example of a failure case without this PR:
```python
import asyncio
from alt_pytest_asyncio.plugin import OverrideLoop

info2 = []


async def my_generator(info):
    try:
        info.append(1)
        yield
        info.append(2)
        yield
        info.append(3)
    except asyncio.CancelledError:
        info.append("cancelled")
        raise
    finally:
        info.append(("done", __import__("sys").exc_info()[0]))


ag = my_generator(info2)


async def doit():
    await ag.__anext__()
    await ag.__anext__()


with OverrideLoop(new_loop=True) as custom_loop:
    custom_loop.run_until_complete(doit())

assert info2 == [1, 2, "cancelled", ("done", asyncio.CancelledError)]
```

The key here is defining `ag = my_generator(info2)` outside. The reason is that asyncio uses `sys.set_asyncgen_hooks()` to set a finalizer that will `athrow()` any async generator _when they are GC'd_. Because there are no references to `ag` after `doit()` is done, it will probably get garbage collected (but this is not guaranteed), and then asyncio's finalizer will create a task to do `athrow()` on the async generator. This task will then get cancelled by `cancel_all_tasks()`, causing the test to pass without `os.kill()`.

In other words, the test may fail occasionally without `os.kill()`, but this is up to GC. By forcing `ag` to remain alive we can get it to fail every time.

Reply:
on looking at that I'm confused why it wouldn't complain that the generator is being created with a different loop to the one you're running. But I suppose that firstiterhook doesn't get called till the first `__anext__` so it could make sense that making the async generator object itself doesn't do anything with the loop.

I'll have a play later. If that removes the os.kill that'd be nice.

Reply:
>  asyncio.run(g.__anext__())
> 1
> asyncio.run(g.__anext__())
> 2

Ever since 2016 for me has been a constant fight with asyncio with because of this project https://github.com/delfick/photons (and specifically the non open source version it's based off)

I'm a big fan of asyncio, but it has some weird edges that cause annoying surprises. The above surprises me because of how angry it gets when you have a future from one loop used by another. So I imagine it'd be fairly simple to make a generator where it complains about the above.

Reply:
Yes, this solves the problem that led to #10. Thanks!

Reply:
wicked awesome! I assume I'll make a release next week. (and if #9 isn't ready by then I'll make a release without that)